lib/srv/db/common/role/role.go

```diff
@@ -17,14 +17,39 @@ limitations under the License.
 package role
 
 import (
+	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/services"
 )
 
-// DatabaseRoleMatchers returns role matchers based on the database protocol.
-func DatabaseRoleMatchers(dbProtocol string, user, database string) services.RoleMatchers {
+// DatabaseRoleMatchers returns role matchers based on the database.
+func DatabaseRoleMatchers(db types.Database, user, database string) services.RoleMatchers {
+	roleMatchers := services.RoleMatchers{
+		services.NewDatabaseUserMatcher(db, user),
+	}
+
+	if matcher, ok := databaseNameMatcher(db.GetProtocol(), database); ok {
+		roleMatchers = append(roleMatchers, matcher)
+	}
+	return roleMatchers
+}
+
+// RequireDatabaseUserMatcher returns true if databases with provided protocol
+// require database users.
+func RequireDatabaseUserMatcher(protocol string) bool {
+	return true // Always required.
+}
+
+// RequireDatabaseNameMatcher returns true if databases with provided protocol
+// require database names.
+func RequireDatabaseNameMatcher(protocol string) bool {
+	_, ok := databaseNameMatcher(protocol, "")
+	return ok
+}
+
+func databaseNameMatcher(dbProtocol, database string) (*services.DatabaseNameMatcher, bool) {
```

Comment:
Nit: can we check for `nil` instead of having an additional `bool` as a parameter?